### PR TITLE
IPS-1109: Canary deployments

### PR DESCRIPTION
**Pull request by louiseli1**
### What changed

Enabled canary deployments for JourneyEngineStepFunction and the following lambdas:

1. IssueClientAccessTokenFunction - `/token` (external)
2. InitialiseIpvSessionFunction - `/session/initialise` (internal)
3. ProcessCriCallbackFunction - `/cri/callback` (internal)
4. BuildUserIdentityFunction - `/user-identity` (external)
5. BuildProvenUserIdentityDetailsFunction - `/user/proven-identity-details`(internal)
6. ProcessAsyncCriCredentialFunction

Lambdas each have alarms for:
- A lambda error
- 5xx error (1-6 only)

These alert #di-ipv-core-non-prod-alarms and #di-ipv-core-prod-alarms

Add DeploymentStrategies:
- JourneyEngineStepFunction to use Versions of lambda, rather than the alias
- Update the API spec to point to the step function alias

### Why did it change

- Canary deployments across identity

### Issue tracking

- [IPS-1109](https://govukverify.atlassian.net/browse/IPS-1109)

#### Canaries in progress
1-CanariesInProgress

#### Starting a failed execution
2-StartFailedExecution

#### Alarm triggered
3-AlarmTriggered

#### Canary reverted to V1
4-CanaryRevertToV1

#### Stack rollback
5-StackRollback

#### Next deployment - canaries in progress
6-CanariesInProgress

#### Canaries complete
7-CanariesComplete

:journey:next

### Lambda 
LambdaDeployment

### Alarms
image

### Environment variables or secrets

- [ ] No environment variables or secrets were added or changed

- [ ] Documented in the [README](./blob/main/README.md)
- [ ] Added to deployment repository
- [ ] Added to local startup repository

### Other considerations

- [x] Update [README](./blob/main/README.md) with any new instructions or tasks

[IPS-1109]: https://govukverify.atlassian.net/browse/IPS-1109?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: deploy/template.yaml [hunk 1]
```diff
@@ -89,6 +93,24 @@ Parameters:
...
+    AllowedValues:
+      - AllAtOnce
+      - Canary10Percent5Minutes
+      - Canary10Percent10Minutes
+      - Canary10Percent15Minutes
+      - Canary10Percent30Minutes
+  StepFunctionsDeploymentPreference:
+    Description: "Specifies the configuration to enable gradual StepFunction deployment."
+    Type: String
+    Default: "CANARY"
```
Comment (sandragds) [inline deploy/template.yaml]: Shall we have default as ALL_AT_ONCE? I think that's the configuration for other services

File: deploy/template.yaml [hunk 2]
```diff
@@ -1247,19 +1297,31 @@ Resources:
...
+        CallTicfCriLambdaArn: !Ref CallTicfCriFunction.Version
+        CallDcmawAsyncCriLambdaArn: !Ref CallDcmawAsyncCriFunction.Version
+        StoreIdentityLambdaArn: !Ref StoreIdentityFunction.Version
+        ResetSessionIdentityFunctionArn: !Ref ResetSessionIdentityFunction.Version
+        CheckCoiFunctionArn: !Ref CheckCoiFunction.Version
+      AutoPublishAlias: live
+      DeploymentPreference:
+        Type: !Ref StepFunctionsDeploymentPreference
+        Interval: 10
+        Percentage: 20
```
Comment (sandragds) [inline deploy/template.yaml]: has the canary strategy been agreed with core?

File: deploy/template.yaml [hunk 3]
```diff
@@ -3003,6 +3101,509 @@ Resources:
...
+  ####################################################################
+  #                                                                  #
+  # Canary Alarms                                                    #
+  #                                                                  #
+  ####################################################################
+
+  JourneyEngineStepFunctionExecutionFailedCanaryAlarm:
+    Type: AWS::CloudWatch::Alarm
+    Properties:
+      ActionsEnabled: true
```
Comment (sandragds) [inline deploy/template.yaml]: start with false so they don't fire to slack for production where it is set to not all at once?

File: deploy/template.yaml [hunk 4]
```diff
@@ -1247,19 +1297,31 @@ Resources:
...
+        CheckExistingIdentityFunctionArn: !Ref CheckExistingIdentityFunction.Version
+        BuildCriOauthRequestFunctionArn: !Ref BuildCriOauthRequestFunction.Version
+        BuildClientOauthResponseFunctionArn: !Ref BuildClientOauthResponseFunction.Version
+        CheckGpg45ScoreFunctionArn: !Ref CheckGpg45ScoreFunction.Version
+        EvaluateGpg45ScoresFunctionArn: !Ref EvaluateGpg45ScoresFunction.Version
+        CallTicfCriLambdaArn: !Ref CallTicfCriFunction.Version
+        CallDcmawAsyncCriLambdaArn: !Ref CallDcmawAsyncCriFunction.Version
+        StoreIdentityLambdaArn: !Ref StoreIdentityFunction.Version
+        ResetSessionIdentityFunctionArn: !Ref ResetSessionIdentityFunction.Version
+        CheckCoiFunctionArn: !Ref CheckCoiFunction.Version
```
Comment (sandragds) [inline deploy/template.yaml:1326]: alarms for all the above functions should be added and configured in the state machine as they are associated with a state machine e.g. in bravo - https://github.com/govuk-one-login/di-id-reuse-storage/blob/3d040e9f94162790a1ad57d0ee862e2a44207901/src/infra/main/template.yaml#L587

File: deploy/template.yaml (hunk 2 above)
Comment (louiseli1) [reply deploy/template.yaml]: Not yet tbc

Comment (louiseli1) [reply deploy/template.yaml]: Yep, strategy updated, we will deploy canaries to build first

**Comment by sandragds** [comment]
These lambdas are without canaries set up and not associated to state machines:
UserReverificationFunction
ReportUserIdentityFunction
BulkMigrateVcsFunction
Should they be set up with canaries too or they dont need them either?

**Comment by louiseli1** [comment]
> These lambdas are without canaries set up and not associated to state machines: UserReverificationFunction ReportUserIdentityFunction BulkMigrateVcsFunction Should they be set up with canaries too or they dont need them either?

They were identified as non critical lambdas, they receive very low traffic and it doesn't seem like canaries would add much benefit

**Review by sandragds — DISMISSED** [review]
Looks good!

File: deploy/template.yaml [hunk 5]
```diff
@@ -3004,6 +3101,728 @@ Resources:
...
+          Value: POST
+        - Name: Stage
+          Value: !Ref Environment
+        - Name: Resource
+          Value: /token
+      Statistic: Sum
+      Unit: Count
+      Period: 60
+      EvaluationPeriods: 1
+      Threshold: 1
```
Comment (Joe-Edwards-GDS) [inline deploy/template.yaml:3181]: Are these deliberately stricter than the error alarms?

File: openAPI/core-back-internal.yaml [hunk 6]
```diff
@@ -75,7 +75,7 @@ paths:
               \"clientOAuthSessionId\":\"$util.escapeJavaScript($input.params('client-session-id'))\",
               \"language\":\"$util.escapeJavaScript($input.params('language'))\"
               }",
-              "stateMachineArn": "arn:aws:states:${AWS::Region}:${AWS::AccountId}:stateMachine:${JourneyEngineStepFunction.Name}"
+              "stateMachineArn": "arn:aws:states:${AWS::Region}:${AWS::AccountId}:stateMachine:${JourneyEngineStepFunction.Name}:live"
```
Comment (Joe-Edwards-GDS) [inline openAPI/core-back-internal.yaml:78]: Do we need to worry about any permissions issues with this change? E.g. I know when we switched to invoking CIMIT lambdas using an alias we had to update our permissions to specify the alias as well.

**Review by Joe-Edwards-GDS — COMMENTED** [review]
Looks reasonable to me, but a couple of queries.

To elaborate on the 'non-critical' lambdas:

*UserReverificationFunction* - This is only used for MFA reset journeys, which is not yet live - but once this journey is live we probably should have canaries for it... (May be several months!)
*ReportUserIdentityFunction* and *BulkMigrateVcsFunction* - These are manually invoked lambdas, used for the bulk EVCS migration, so it doesn't make sense to canary these.

Goes without saying, but please don't merge until next week!

File: openAPI/core-back-internal.yaml (hunk 6 above)
Comment (louiseli1) [reply openAPI/core-back-internal.yaml:78]: I’ve looked at 3 types of permissions:

1. APIGW to step function with `JourneyEngineStepFunctionApiGateWayIamRole`: No changes needed to this, APIGW can invoke any version of the live alias of the state machine with a response 200

2. Step function to lambdas with LambdaInvokePolicy: Also no changes needed, this is an inbuilt policy and gives permission to invoke a function, alias, or version (https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-policy-template-list.html#lambda-invoke-policy)

3. Test APIGW to lambdas: It looks like the APIGW already has access to `Function.Alias` https://github.com/govuk-one-login/ipv-core-back/blob/eaf5cd912f91798148f77427ef6b5b5c75c7c38e/deploy/template.yaml#L858 - but do you know how I could double check this works for both versions during a canary deployment? Ie what query strings/headers do I need to test permissions from apigw to these lambdas for build-user-identity, initialise-ipv-session, issue-client-access-token etc

Is there anything else I should check? I’ve used https://orch-dev-sre.01.core.dev.stubs.account.gov.uk/ to run through journeys (not sure if I’m doing this bit right?), and it looks like the canaries are splitting traffic, see screenshot from when I upped the ratio from 10:90 to 50:50):

step-function-core-canaries

File: deploy/template.yaml (hunk 5 above)
Comment (louiseli1) [reply deploy/template.yaml:3181]: Not intentional, there is a baseline error rate so I've modified these to also be in 3x1 minute, same for the state machine `ExecutionsFailed` alarm E.g. this is the sum over 5 min periods in build over the last 4 weeks:

executions-failed-core-build

**Comment by louiseli1** [comment]
> Looks reasonable to me, but a couple of queries.
> 
> To elaborate on the 'non-critical' lambdas:
> 
> _UserReverificationFunction_ - This is only used for MFA reset journeys, which is not yet live - but once this journey is live we probably should have canaries for it... (May be several months!) _ReportUserIdentityFunction_ and _BulkMigrateVcsFunction_ - These are manually invoked lambdas, used for the bulk EVCS migration, so it doesn't make sense to canary these.
> 
> Goes without saying, but please don't merge until next week!

I’ve updated the PR description to remove the manual lambdas, and have taken note to add UserReverificationFunction

File: deploy/template.yaml (hunk 5 above)
Comment (Joe-Edwards-GDS) [reply deploy/template.yaml:3181]: Thanks - that's a higher baseline than I'd have expected tbh, so maybe something for me to look into!

File: openAPI/core-back-internal.yaml (hunk 6 above)
Comment (Joe-Edwards-GDS) [reply openAPI/core-back-internal.yaml:78]: TBH if you've run through journeys then that's the best test :)

**Review by Joe-Edwards-GDS — APPROVED** [review]

**Pull request merged**